src/encode.js

```diff
@@ -10,13 +10,30 @@ import ParseObject from './ParseObject';
 import { Op } from './ParseOp';
 import ParseRelation from './ParseRelation';
 
+const MAX_RECURSIVE_CALLS = 999;
+let recursiveCallsCount = 0;
```

Comment:
This implementation doesn't work. The counter resides at the module scope and is "global" for the module. Each subsequent call to `encode` uses the same global instance of the variable.

Instead, you could add a counter argument to the `encode` function, so that it has a call-scoped counter.

Reply:
I made the requested argument counter change, but I was struggling to write a test because I don't know much about how Parse works behind the scenes. The `ParseObject` import is mocked in the `encode` test file and I was going down a rabbit hole trying to instantiate the class like I saw in some other test files.

Reply:
You could just write a test for the encode function directly. Maybe a new test file for `encode.js`. Then just run it with dummy args so that it runs recursively. But writing a higher level test would be better, I agree. For that you could look at in what cases `encode` runs recursively, then write something that makes it run recursively. I'm just guessing but maybe an object that references itself would cause a recursing. Maybe something like a circular Parse Pointer, where pointer A references pointer B which references pointer A... any maybe that is even what is causing the issue you are observing in our deployment.

Reply:
I can't get it to work whatsoever, this is what I have:

```js
it('handles circular references in ParseObjects', () => {
  const RealParseObject = jest.requireActual('../ParseObject').default;
  const obj1 = new RealParseObject();
  const obj2 = new RealParseObject();
  obj1.set('obj2', obj2);
  obj2.set('obj1', obj1);

  expect(() => encode(obj1, false, false, [], false)).toThrow();
});
```

This throws a TypeError that has nothing to do with my code:

```shell
TypeError: Cannot read properties of undefined (reading 'estimateAttributes')

      153 |   get attributes(): AttributeMap {
      154 |     const stateController = CoreManager.getObjectStateController();
    > 155 |     return Object.freeze(stateController.estimateAttributes(this._getStateIdentifier()));
          |                                          ^
      156 |   }
      157 |
      158 |   /**

      at ParseObject.estimateAttributes [as attributes] (src/ParseObject.js:155:42)
      at ParseObject.attributes [as set] (src/ParseObject.js:747:36)
      at Object.set (src/__tests__/encode-test.js:192:10)
```

And if I try including `obj1.save()` before setting anything, I get this error:

```shell
TypeError: Cannot read properties of undefined (reading 'save')

      1333 |     const controller = CoreManager.getObjectController();
      1334 |     const unsaved = options.cascadeSave !== false ? unsavedChildren(this) : null;
    > 1335 |     return controller.save(unsaved, saveOptions).then(() => {
           |                       ^
      1336 |       return controller.save(this, saveOptions);
      1337 |     });
      1338 |   }

      at ParseObject.save (src/ParseObject.js:1335:23)
      at Object.save (src/__tests__/encode-test.js:192:10)
```

I have no idea how to proceed from here, so if anyone else knows how this library works and can implement a test, that would be great.

Reply:
Are you able to run the existing tests locally without issues? Also, when you make any changes, in /src, you need to rebuild the SDK, either with `npm run build` or continuously rebuild on file change with `npm run watch`.

Reply:
Yes, I can run the existing tests without issues. And re-running `npm run build` did not make a difference to the tests, this test:

```js
it('handles circular references in ParseObjects', () => {
  const RealParseObject = jest.requireActual('../ParseObject').default;
  const obj1 = new RealParseObject();
  const obj2 = new RealParseObject();
  obj1.set('obj2', obj2);
  obj2.set('obj1', obj1);

  expect(() => encode(obj1, false, false, [], false)).toThrow();
});
```

Still gives this error:

```js
TypeError: Cannot read properties of undefined (reading 'estimateAttributes')

      153 |   get attributes(): AttributeMap {
      154 |     const stateController = CoreManager.getObjectStateController();
    > 155 |     return Object.freeze(stateController.estimateAttributes(this._getStateIdentifier()));
          |                                          ^
      156 |   }
      157 |
      158 |   /**

      at ParseObject.estimateAttributes [as attributes] (src/ParseObject.js:155:42)
      at ParseObject.attributes [as set] (src/ParseObject.js:747:36)
      at Object.set (src/__tests__/encode-test.js:192:10)
```

Reply:
How about the test `encodes ParseObjects` which just encodes a regular obj?

Reply:
See https://github.com/parse-community/Parse-SDK-JS/issues/2098#issuecomment-2027459521 for how you may reproduce this issue.